tool/tsh/common/tsh_test.go

```diff
@@ -1773,231 +1773,269 @@ func (o *output) String() string {
 // ssh server using a resource access request when "tsh ssh" fails with
 // AccessDenied.
 func TestSSHAccessRequest(t *testing.T) {
-	modules.SetTestModules(t, &modules.TestModules{TestBuildType: modules.BuildEnterprise})
-	tmpHomePath := t.TempDir()
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	requester, err := types.NewRole("requester", types.RoleSpecV6{
-		Allow: types.RoleConditions{
-			Request: &types.AccessRequestConditions{
-				SearchAsRoles: []string{"node-access"},
-			},
+	tests := []struct {
+		name        string
+		requestMode string
+	}{
+		{
+			name:        "resource-based",
+			requestMode: accessRequestModeResource,
 		},
-	})
-	require.NoError(t, err)
-
-	nodeAccessRole, err := types.NewRole("node-access", types.RoleSpecV6{
-		Allow: types.RoleConditions{
-			NodeLabels: types.Labels{
-				"access": {"true"},
-			},
-			Logins: []string{"{{internal.logins}}"},
+		{
+			name:        "role-based",
+			requestMode: accessRequestModeRole,
 		},
-	})
-	require.NoError(t, err)
-
-	connector := mockConnector(t)
-
-	alice, err := types.NewUser("alice@example.com")
-	require.NoError(t, err)
-	alice.SetRoles([]string{"requester"})
-	user, err := user.Current()
-	require.NoError(t, err)
-	traits := map[string][]string{
-		constants.TraitLogins: {user.Username},
 	}
-	alice.SetTraits(traits)
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
 
-	rootAuth, rootProxy := makeTestServers(t,
-		withBootstrap(requester, nodeAccessRole, connector, alice),
-		withConfig(func(cfg *servicecfg.Config) {
-			cfg.Clock = clockwork.NewFakeClock()
-		}))
+			modules.SetTestModules(t, &modules.TestModules{TestBuildType: modules.BuildEnterprise})
+			tmpHomePath := t.TempDir()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			requester, err := types.NewRole("requester", types.RoleSpecV6{
+				Allow: types.RoleConditions{
+					Request: &types.AccessRequestConditions{
+						Roles:         []string{"node-access"},
+						SearchAsRoles: []string{"node-access"},
```

Comment:
I'd love to see some testcases where `request.roles` and `request.search_as_roles` don't match. Will we offer an access request when `search_as_roles` lets us list the resource but none of the `roles` actually grant access?